### ISSUE TITLE
Notification on startup that multi-threading is in use

### ISSUE THREAD

**Issue by revodavid** [RevolutionAnalytics/RRO#134]
On startup, issue a message to indicate multu-threaded math is in use:

`Multithreaded BLAS/LAPACK libraries detected. [NN] cores will be used.`

On MKL platforms (Windows/Linux), [NN] should be the results of `RevoMathUtils::getMKLthreads()`. On Mac, simply print `All` or count the number of cores some other way.

**Comment by andrie** [comment]
Agree with this.  We discussed this last night in a team meeting, and agreed to implement in 8.0.4.

Note for implementation:

+ If `RevoUtilsMath` is installed, display message about number of cores used
+ If not installed, display message about optional download instructions

**Comment by sdgilley** [comment]
I agree it should be for 8.0.4, we still need to work out some of the details…
For example, David S. wants a message on Mac as well.  But Mac does not currently install RevoUtilsMath (or is is now RevoMathUtils?  I keep getting confused…) but that is not installed today on the Mac.  Should we install it in 8.0.4?

Let’s discuss more next week…

From: Andrie de Vries
Reply-To: RevolutionAnalytics/RRO
Date: Thursday, April 30, 2015 at 10:09 AM
To: RevolutionAnalytics/RRO
Subject: Re: [RRO] Notification on startup that multi-threading is in use (#134)


Agree with this. We discussed this last night in a team meeting, and agreed to implement in 8.0.4.

Note for implementation:

  *   If RevoUtilsMath is installed, display message about number of cores used
  *   If not installed, display message about optional download instructions

**Comment by andrie** [comment]
On windows and Linux, the code is:

```r
if(require("RevoUtilsMath")){
  local({
    ncores <- RevoUtilsMath::getMKLthreads()
    msg <- "Multithreaded BLAS/LAPACK libraries detected. Using %d cores for math algorithms."
    sprintf(msg, ncores)
  })
} 
```

However, I need to find a reliable way of testing for the availability of math library on Mac.

@revodavid and @sdgilley Do you know how to do this?

**Comment by revodavid** [comment]
Since our Mac builds are always linked to Accelerate, there's no need to detect. The message could thus be:

"Multithreaded BLAS/LAPACK libraries detected. Using all available cores for math algorithms."

You could use parallel:::detectCores() to actually count the number, but at least on my MBP that reports 8 instead of the expected 4.

# David

**Comment by andrie** [comment]
I see what you mean, but this implies a forked version of `Rprofile.site`, which makes maintenance harder.  IIUC, right now we have only a single version of `Rprofile.site`.

**Comment by revodavid** [comment]
You could have platform check in Rprofile.site with branch code specific to platforms, without needing to have multiple versions of the file.

**Comment by andrie** [comment]
Yes, but I haven't found a way to reliably tell me that the platform is OSX.  The OS checks I found tells me that the Mac OS is "Linux", and testing for the package type, e.g. "mac.binary" is explicitly recommended against. In any way, this isn't guaranteed to work, since the setting could be "source", even on a Mac.

**Comment by revodavid** [comment]
`Sys.info()["sysname"]=="Darwin"` should be an accurate test for Mac platform on all the Mac versions we run on.

**Comment by andrie** [comment]
This code seems to work on Windows and Mac. Not yet tested on Linux:

```r
if(Sys.info()["sysname"] == "Darwin"){
  local({
    hw.ncpu <- system('sysctl hw.ncpu', intern = TRUE)
    ncores <- sub("hw.ncpu: ", "", hw.ncpu)
    msg <- "Multithreaded BLAS/LAPACK libraries detected. Using %s cores for math algorithms."
    message(sprintf(msg, ncores))
    invisible(NULL)
  })
} else if(require("RevoUtilsMath")){
  local({
    ncores <- RevoUtilsMath::getMKLthreads()
    msg <- "Multithreaded BLAS/LAPACK libraries detected. Using %d cores for math algorithms."
    message(sprintf(msg, ncores))
    invisible(NULL)
  })
} 
```

**Comment by j-martens** [comment]
Yes, it worked on my Mac.

Do we need to verify on Linux? Who can do that?

**Comment by revodavid** [comment]
Deferred to 3.2.1

**Comment by j-martens** [comment]
Next step needed is the error message. When you are ready for text, @nathansoz , please ask @j-martens for input.

**Comment by nathansoz** [comment]
I would like to propose a minor modification. The below code checks to see if RevoUtilsMath exists. If it doesn't exist, then it silently doesn't load RevoUtilsMath. If it does exist, then it uses the code @andrie wrote. Thoughts?

``` R
load_if_installed <- function(package) {
  if(package %in% rownames(utils::installed.packages())) {
    do.call('library', list(package))
    return(TRUE)
  }
  else
    return(FALSE)

}

if(Sys.info()["sysname"] == "Darwin"){
  local({
    hw.ncpu <- system('sysctl hw.ncpu', intern = TRUE)
    ncores <- sub("hw.ncpu: ", "", hw.ncpu)
    msg <- "Multithreaded BLAS/LAPACK libraries detected. Using %s cores for math algorithms."
    message(sprintf(msg, ncores))
    invisible(NULL)
  })
} else if(load_if_installed("RevoUtilsMath")){
  local({
    ncores <- RevoUtilsMath::getMKLthreads()
    msg <- "Multithreaded BLAS/LAPACK libraries detected. Using %d cores for math algorithms."
    message(sprintf(msg, ncores))
    invisible(NULL)
  })
}
```

**Comment by richcalaway** [comment]
installed.packages() can be a lot of overhead--it creates a (potentially huge) data frame with a bunch of info about the installed packages. A simpler check involves looking for the packages DESCRIPTION file:

if (!identical(system.file(package="RevoUtilsMath"), ""))

**Comment by richcalaway** [comment]
I don't see the second part of the request:

•If not installed, display message about optional download instructions

**Comment by nathansoz** [comment]
@j-martens and I discussed this and thought it might be slightly overbearing to print that message each time the user starts R. She was going to check in with @revodavid about whether it was needed.

If we do need it, we agreed it can be added with a separate pull request.

**Comment by richcalaway** [comment]
One more afterthought...we should probably move this whole batch of message code inside the if (!quiet) conditional, so that if users are trying to suppress startup messages they won't see this one.

**Comment by sfweller** [comment]
Nathan, your proposed R function looks good to me, but I'm not sure when we would be running/checking this code?  Would this be something that the installer runs at install time or would this be checked when launching the RRO?

**Comment by nathansoz** [comment]
@richcalaway Wouldn't that make it so that people who disable startup messages wouldn't have RevoUtilsMath loaded on startup?

**Comment by richcalaway** [comment]
True enough, but the two items are fully separable. You can load the package without printing the message. Something like this should work

local({
  
  if(Sys.info()["sysname"] == "Darwin"){
    hw.ncpu <- system('sysctl hw.ncpu', intern = TRUE)
    ncores <- as.numeric(sub("hw.ncpu: ", "", hw.ncpu))
    msg <- "Multithreaded BLAS/LAPACK libraries detected. Using %d cores for math algorithms."    
  }
  else if(load_if_installed("RevoUtilsMath")){
    ncores <- RevoUtilsMath::getMKLthreads()
    msg <- "Multithreaded BLAS/LAPACK libraries detected. Using %d cores for math algorithms."
  }
  else {
	ncores <- 1
	msg <- "No multithreaded BLAS/LAPACK libraries detected. Using %d cores for math algorithms."
  }
	if (!quiet) {
		message(sprintf(msg, ncores))
	}
}
)

**Comment by nathansoz** [comment]
Hey Rich, I took your suggestion with a small modification. My reading of the docs on local() seems to make me think that things called within it don't persist. So the loading of RevoUtilsMath wouldn't persist in the code you wrote above... If this isn't the case let me know! Otherwise, I wrote some code the preserves the separate local() blocks that @andrie wrote above but also suppresses console output if the user passes quiet flags.
